### ISSUE TITLE
document 3.5.0 namespace-aware hooks change as potentially breaking

### ISSUE THREAD

**Issue by dreyks** [capistrano/capistrano#1652]
one of my cap files

```ruby
namespace :app do
  task :restart do
    # restarts the app server
  end

  after 'deploy:publishing', :restart
end
```

After upgrading to 3.5.0 my deploy silently ignored the `after` directive.
Had to change it to `after 'deploy:publishing', 'app:restart'`

**Comment by leehambley** [comment]
Are you referring to this change from the changelog?

> Allow after() to refer to tasks that have not been loaded yet (@jcoglan)

**Comment by dreyks** [comment]
Nope. This one. Under `Other changes`

> Ensure task invocation within after hooks is namespace aware

**Comment by dreyks** [comment]
However yes, my situation is an accumulation of both of this changes:
1. :restart task was being searched in global namespace
2. no exception was raised on not finding it though because `after` can refer to not yet loaded tasks

As a result the task was silently not run and the whole deploy reported as if everything is ok

**Comment by mattbrictson** [comment]
> Ensure task invocation within after hooks is namespace aware

Yes, this is the underlying problem, and we should document it as a potentially breaking change.

I believe this is actually what is happening:

1. By default, Capistrano defines a `deploy:restart` task. This task is empty. ([source](https://github.com/capistrano/capistrano/blob/44af53564beede8ba236b3f97d138e39d5b3cd20/lib/capistrano/tasks/deploy.rake#L237))
2. Your `after` declaration means that after `deploy:publishing`, Capistrano searches for a task named `:restart`.
3. In Capistrano 3.5.0, Capistrano first searches the `deploy` namespace.
4. It finds the existing `deploy:restart` task and invokes it.
5. Since the `deploy:restart` task is empty by default, nothing happens.

If a task doesn't exist at invocation time, you will still get an error in Capistrano 3.5.0. This change:

> Allow after() to refer to tasks that have not been loaded yet

still means the task must ultimately be defined somewhere, else an error will occur. It simply allows the definition to be deferred.

So, to summarize, your solution is correct:

> Had to change it to `after 'deploy:publishing', 'app:restart'`

but I thought I should explain further the mechanism of why it works.

**Comment by mattbrictson** [comment]
Original PR that introduced this change (for future reference): #1516

**Comment by dreyks** [comment]
Oh I see. The default `restart` task was my first thought too, but then I saw this change and thought that it was related.

Anyway, this may be closed

**Comment by mattbrictson** [comment]
Will close automatically when #1653 is merged. Thanks for your report!

**Comment by thickpaddy** [comment]
Hi,

Sorry, this issue is an unintentional side effect of my naive fix to make the post task lookup for after hooks namespace aware again, after another change to lazily bind the post task that affected this behaviour.

My code looks up the post task using the namespace of the original task, assuming that the current namespace matches the scope of the original task, but this isn't always true, as outlined in this issue report.

This issue can be fixed with a minor change to the code, to use `Rake.application.current_scope` rather than `task.scope` when looking up the post task.

I'll submit another pull request asap.

Mark

**Comment by mattbrictson** [comment]
@thickpaddy Please see existing issue #1543 on that exact topic, and my PR #1547 which I believe does the same fix that you just mentioned. As you can see in that PR I was not able to merge the change because it causes prominent Capistrano plugins to break.

My conclusion was that `task.scope` may not be the *correct* way to look up the post task, but that ship has already sailed.

**Comment by thickpaddy** [comment]
Thanks @mattbrictson, you've explained it in great detail, and yeah, it seems my quick fix to the post task lookup is the same as yours, which doesn't work with some plugins as you've pointed out :-(

The post task lookup in after hooks should at least be consistent with pre task lookup in before hooks, even if that doesn't match rake's behaviour, and my previous change broke that consistency. It would be nice to at least address that inconsistency if possible. I'm not sure how, but I'd like to take a look when I get some time and see if it can be done independent of addressing the wider issue of inconsistency between rake and capistrano explained in #1547

**Comment by mattbrictson** [comment]
@thickpaddy Yes, I agree it may be possible to tweak the `after` behavior without breaking things. It looks like the trouble I was encountering was mostly regarding the `before` hooks.
